src/components/select/CustomSelect.tsx

```diff
@@ -0,0 +1,259 @@
+  const dropdownRect = dropdown.getBoundingClientRect();
+  const bottomToViewportBottom = window.innerHeight - dropdownRect.bottom;
+  const dropdownHeight = dropdown?.offsetHeight || 0;
+
+  // If the dropdown is cut off at the bottom of the viewport
+  // adjust the height to fit within the viewport minus fixed margin.
+  // This usually becomes an issue when the dropdown is at the bottom of the viewport or screen getting smaller.
+  if (bottomToViewportBottom < 0) {
+    const adjustedHeight = dropdownHeight + bottomToViewportBottom - 20;
+    dropdown.style.height = `${adjustedHeight}px`;
+    dropdown.style.maxHeight = `${adjustedHeight}px`;
+    return;
+  }
+
+  // If the dropdown is taller than the max height, do not adjust.
+  if (dropdownHeight >= DROPDOWN_MAX_HEIGHT) {
+    return;
+  }
+
+  // If the dropdown does not have overflow, do not adjust.
+  const hasOverflow = dropdown.scrollHeight > dropdown.clientHeight;
+  if (!hasOverflow) {
+    return;
+  }
+
+  // If the dropdown is not cut off at the bottom of the viewport
+  // adjust the height of the dropdown so that its bottom edge is 20px from the bottom of the viewport.
+  const topToViewportBottom = window.innerHeight - dropdownRect.top;
+  const adjustedHeight = topToViewportBottom - 20;
+  dropdown.style.height = `${adjustedHeight}px`;
+  dropdown.style.maxHeight = `${adjustedHeight}px`;
+};
+
+const adjustDropdownHeightAbove = (dropdown: HTMLElement) => {
+  // The search height is subtracted (if necessary) so that no options will be hidden behind the search input.
+  const search = document.querySelector(
+    ".p-custom-select__search",
+  ) as HTMLElement;
+  const searchRect = search?.getBoundingClientRect();
+  const searchHeight = searchRect?.height || 0;
+  const dropdownRect = dropdown.getBoundingClientRect();
+  const bottomToViewportTop = dropdownRect.bottom;
+  const dropdownHeight = dropdown?.offsetHeight || 0;
+
+  // If the dropdown is taller than the top edge of the viewport
+  // adjust the height to fit within the viewport minus fixed margin.
+  if (dropdownHeight > bottomToViewportTop) {
+    const adjustedHeight = bottomToViewportTop - searchHeight - 20;
+    dropdown.style.height = `${adjustedHeight}px`;
+    dropdown.style.maxHeight = `${adjustedHeight}px`;
+    return;
+  }
+
+  // If the dropdown is taller than the max height, do not adjust.
+  if (dropdownHeight >= DROPDOWN_MAX_HEIGHT) {
+    return;
+  }
+
+  // If the dropdown does not have overflow, do not adjust.
+  const hasOverflow = dropdown.scrollHeight > dropdown.clientHeight;
+  if (!hasOverflow) {
+    return;
+  }
+
+  // If the dropdown is not cut off at the top of the viewport
+  // adjust the height of the dropdown so that its top edge is 20px from the top of the viewport.
+  const adjustedHeight = bottomToViewportTop - searchHeight - 20;
+  dropdown.style.height = `${adjustedHeight}px`;
+  dropdown.style.maxHeight = `${adjustedHeight}px`;
+};
```

Comment:
I've moved the functions into a `customSelect.tsx` utils file and created some test. Also I've cleaned up the functions a bit  to make it more readable. Although I still prefer having the above functions separate since a lot of the logical comparison as well as the offset calculations depend on the dropdown position. Creating a single function would mean we have to check the position of the dropdown and then decide on how to do the calculations, I think it would make the code more difficult to understand.

Reply:
LGTM as is with the dedicated file and test applied.